### PR TITLE
Initial work on better handling errors

### LINKED ISSUE #645: Add global error catch

### LINKED ISSUE #1209: Preload should support failure (such as 404) and deal appropriately

### LINKED ISSUE #1280: Investigate recovering and logging after errors
Ideally, navigating to a new page would recover from an error. We may also benefit from logging errors.

### LINKED ISSUE #1590: 400 errors have "undefined" in the message
screen shot 2017-04-13 at 4 57 44 pm

### LINKED ISSUE #1948: wire up caught errors with posted stack traces
follow up to #645

### PR DESCRIPTION
**Pull request by eatonphil**
A few things:

* Gets rid of full error pages except for 404s
* Adds a top-level div to handle rendering an emergency modal when no other part of the page can be rendered by the root div. Closes #645 
* Catching errors during `render()` is tricky. See notes in `src/index.js`.
* POSTs logs to /error, closes #1948 
* Full page change after error occurs, closes #1280 
* Closes #1590 
* Closes #1209

To test: `NODE_ENV=production npm start` so red error screen doesn't block our error handling UI. You'll also want to `npm install && npm run rebuild-components`.

TODO:
* [x] Check out GA for logging errors
* [x] Check out IPManagement page for errors and error handling there.

### CODE REVIEW

**Comment by mparke** [comment]
Please look at https://github.com/csnover/TraceKit before continuing

**Comment by na3d** [comment]
Code review wise lgtm.

Resolve the conflicts, then I will pull it down again to go through.

**Comment by eatonphil** [comment]
@na3d done.

**Comment by mparke** [comment]
is this the new 404? might need some css, margin/padding etc

screen shot 2017-06-07 at 4 30 44 pm

**Comment by eatonphil** [comment]
@mparke Make sure you `npm run rebuild-components`.

**Comment by mparke** [comment]
When forcing an error, seeing this in the console:

```
[Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
```

**Comment by eatonphil** [comment]
@mparke it's nothing new we're doing. TraceKit is saying that.

**Comment by mparke** [comment]
so if I purposefully create a bug in a render method right now, I red screen and get the error repeated several times in the console with no other page change,

is this expected?

**Comment by eatonphil** [comment]
You've got to be running `NODE_ENV=production npm start` otherwise those red screens will show up.

File: devServer.js [hunk 1]
```diff
@@ -28,13 +28,19 @@ app.use(require('webpack-dev-middleware')(compiler, {
...
 app.get('*', function(req, res) {
-  if (req.url === "/static/common.js") {
+  if (req.url === '/static/common.js') {
     res.send('');
   } else {
     res.sendFile(path.join(__dirname, 'index.html'));
   }
 });
 
+app.post('/error', function(req, res) {
```
Comment (mparke) [inline devServer.js]: I assume this is going away?

Comment (eatonphil) [inline devServer.js]: Yup, just pushed the removal.

File: src/index.js [hunk 2]
```diff
@@ -1,101 +1,175 @@
...
+        </ModalShell>,
+        document.getElementById('emergency-modal')
+      );
+    }
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.trace(e);
+  }
+
+  // TraceKit.report throws an error.
```
Comment (mparke) [inline src/index.js]: on purpose? or as in also needs to be handled

Comment (eatonphil) [inline src/index.js]: On purpose.

File: src/router.js [hunk 3]
```diff
@@ -46,9 +46,24 @@ export class LoadingRouterContext extends RouterContext {
...
+            error = true;
+            window.handleError(e);
+            this.setState({ noRender: true });
+          }
+
           this.preloadCounter--;
+
+          // If a preload triggered an error, stop preloading and all the error to be shown.
+          if (error) {
+            break;
```
Comment (mparke) [inline src/router.js]: `break;` v. `return;`?

Comment (eatonphil) [inline src/router.js]: Break, this is for the for loop it's in.

File: src/index.js [hunk 4]
```diff
@@ -1,101 +1,175 @@
...
-      <div>
-        <Router
-          history={history}
-          onUpdate={logPageView}
-          dispatch={store.dispatch}
-          render={props => <LoadingRouterContext {...fillInMissingProps(props)} />}
+window.handleError = function (e) {
+  try {
+    // eslint-disable-next-line no-console
+    console.trace(e);
```
Comment (eatonphil) [inline src/index.js]: I don't think there's any harm in leaving it.

**Comment by mparke** [comment]
So, testing this again in ga debug mode, I think I've validated that it's executing correctly:

screen shot 2017-06-08 at 11 27 38 am

I'm guessing there's a delay in processing on GA's side, but i'm not sure.

File: src/router.js (hunk 3 above)
Comment (mparke) [inline src/router.js]: ah... shame this isn't a nice `while` then

**Pull request merged**